phase1-coordinator/src/coordinator_state.rs

```diff
@@ -2323,30 +2417,86 @@ impl CoordinatorState {
         // Fetch the timeout threshold for contributors.
-        let contributor_timeout = self.environment.contributor_timeout_in_minutes() as i64;
+        let participant_lock_timeout = self.environment.participant_lock_timeout();
 
         // Fetch the current time.
-        let now = Utc::now();
+        let now = time.utc_now();
 
-        // Initialize the list of justifications.
-        let mut justifications = vec![];
+        self.current_contributors
+            .clone()
+            .iter()
+            .chain(self.current_verifiers.clone().iter())
+            .filter_map(|(participant, participant_info)| {
+                dbg!(participant);
+                dbg!(&participant_info.locked_chunks);
+                if !self.is_coordinator_contributor(&participant)
+                    && participant_info
+                        .locked_chunks
+                        .values()
+                        .find(|lock| {
+                            let elapsed = now - lock.lock_time;
+                            elapsed > participant_lock_timeout
+                        })
+                        .is_some()
+                {
+                    Some(self.drop_participant(participant, time))
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
 
-        // Process the contributors.
-        for (participant, participant_info) in &self.current_contributors.clone() {
-            // Fetch the elapsed time.
-            let elapsed = now - participant_info.last_seen;
+    /// This will drop a contributor if it hasn't been seen for more
+    /// than [crate::environment::Environment]'s
+    /// `contributor_seen_timeout`.
+    fn update_contributor_seen_drops(&mut self, time: &dyn TimeSource) -> Result<Vec<Justification>, CoordinatorError> {
+        // Fetch the timeout threshold for contributors.
+        let contributor_seen_timeout = self.environment.contributor_seen_timeout();
 
-            // Check if the participant is still live and not a coordinator contributor.
-            if elapsed.num_minutes() > contributor_timeout && !self.is_coordinator_contributor(&participant) {
-                // Drop the participant.
-                justifications.push(self.drop_participant(participant)?);
-            }
-        }
+        // Fetch the current time.
+        let now = time.utc_now();
 
-        Ok(justifications)
+        self.current_contributors
+            .clone()
+            .iter()
+            .filter_map(|(participant, participant_info)| {
+                // Fetch the elapsed time.
+                let elapsed = now - participant_info.last_seen;
+
+                // Check if the participant is still live and not a coordinator contributor.
+                if elapsed > contributor_seen_timeout && !self.is_coordinator_contributor(&participant) {
```

Comment:
This part with a similar place in `update_participant_lock_drops` seems like the main change of logics in this PR, is it?

Reply:
This particular part of the code contains the same logic (with a new name, and some use of iterators). The main change in how it works is actually the contributor's heartbeat thread and the new heartbeat endpoint in the coordinator server. But you're right that the same place in `update_participant_lock_drops` is a significant change to the logic, with a new form of dropping when locks have been held for too long.